### PR TITLE
[TT-Transformers] Support on-device sampling 

### PR DESCRIPTION
**Pull request by sraizada-tt**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/25721

### What's changed
This PR adds support for on-device sampling for TT-Transformers models, extending beyond the previous greedy-only decoding to support top-k, top-p, and temperature-based sampling. The implementation creates a unified TTSampling module that works across different model implementations.

### Key changes

- Introduced a new TTSampling class in `models/common/tt_sampling.py` that provides flexible on-device sampling with support for top-k, top-p, and temperature parameters
- Updated model classes to integrate the sampling module and replaced the `argmax_on_device` parameter with `sampling_on_device` throughout the codebase
- Modified demo configurations to include `top_k` parameter in sampling configurations

### Checklist
Blackhole demo: https://github.com/tenstorrent/tt-metal/actions/runs/18875468902
T3K demo: https://github.com/tenstorrent/tt-metal/actions/runs/18877479493
vllm nightly: https://github.com/tenstorrent/tt-metal/actions/runs/18834427967
Galaxy demo: https://github.com/tenstorrent/tt-metal/actions/runs/18898980448
Single card demo: https://github.com/tenstorrent/tt-metal/actions/runs/18899305656

### CODE REVIEW

**Review by gwangTT — APPROVED** [review]

**Review by handrewsTT — APPROVED** [review]

File: models/tt_transformers/demo/simple_text_demo.py [hunk 1]
```diff
@@ -291,7 +291,11 @@ def prepare_generator_args(
             200,  # max_generated_tokens
             True,  # paged_attention
             {"page_block_size": 32, "page_max_num_blocks_per_dp": 1024},  # page_params
-            {"temperature": 0, "top_p": 0.08},  # sampling_params (argmax)
+            {
+                "temperature": torch.linspace(0.0, 1.0, steps=32).tolist(),
+                "top_p": torch.linspace(0.08, 1.0, steps=32).tolist(),
```
Comment (mtairum) [inline models/tt_transformers/demo/simple_text_demo.py:310]: Why are we using linspace for this test?

File: models/tt_transformers/demo/simple_text_demo.py [hunk 2]
```diff
@@ -856,12 +860,16 @@ def test_demo_text(
...
-        else:
-            device_sampling_params = None
+        # Use device sampling for all cases when supported
+        device_sampling_params = (
+            SamplingParams(
+                temperature=sampling_params["temperature"],
+                top_k=sampling_params["top_k"],
+                top_p=sampling_params["top_p"],
+            )
+            if model_args[0].vocab_size // num_devices <= 64 * 1024 and os.environ.get("MESH_DEVICE") != "N150"
```
Comment (mtairum) [inline models/tt_transformers/demo/simple_text_demo.py]: Please create an issue for sampling on N150 and tag that issue as a comment here.

File: models/tt_transformers/tt/generator.py [hunk 3]
```diff
@@ -252,22 +253,30 @@ def decode_forward_text(
         read_from_device=True,
         sampling_params: SamplingParams = None,  # Should be None if not greedy decoding / sampling on device.
     ):
-        assert (
-            sampling_params is None or sampling_params.temperature == 0
-        ), "Currently only supporting greedy decoding (temperature=0) on device"
-        argmax_on_device = sampling_params is not None and sampling_params.temperature == 0
+        # Use on-device sampling when sampling params are provided
+        if self.mesh_device.shape == (1, 1):
+            assert sampling_params is None, "Sampling on device is not supported on N150"
```
Comment (mtairum) [inline models/tt_transformers/tt/generator.py]: Please create a sampling N150 issue and comment with the issue number this section of the code.
Also the text should be expanded to included P150 as well.

File: models/tt_transformers/tt/model.py [hunk 4]
```diff
@@ -129,6 +130,17 @@ def __init__(
             max_columns_per_device=self.args.max_columns_per_device_lm_head,
         )
 
+        # Initialize on-device sampling if supported
+        # N150 does not support on-device sampling
+        # Sampling on device is supported only if each device has maximum logits size of 64*1024
```
Comment (mtairum) [inline models/tt_transformers/tt/model.py]: Please create a ticket for this issue and add to this comment (N150 and P150)

Comment (mtairum) [reply models/tt_transformers/tt/model.py]: I'm assuming this is the same uint16 limitation..

File: models/tt_transformers/tt/model.py [hunk 5]
```diff
@@ -321,8 +333,9 @@ def process_output_decode(self, tt_out, B, S=1, is_tokens=False):
         Input is ttnn host tensor of logits if is_tokens=False, otherwise tokens. Output is the corresponding torch tensor.
         """
         if is_tokens:
+            padded_batch_size = 32
+            tt_out = ttnn.reshape(tt_out, ttnn.Shape([1, 1, padded_batch_size, 1]))
```
Comment (mtairum) [inline models/tt_transformers/tt/model.py:357]: Why are we padding to 32?
Add a comment.

**Review by mtairum — CHANGES_REQUESTED** [review]

File: models/common/tt_sampling.py [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+
+# SPDX-License-Identifier: Apache-2.0
+
+import logging
+from dataclasses import fields, replace
+from typing import List
+
+import torch
+
+import ttnn
```
Comment (mtairum) [inline models/common/tt_sampling.py:11]: Remove extra spacing.

File: models/common/tt_sampling.py [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+        sub_core_grids: Sub-core grid configuration for operations
+        sub_core_grid_topk: Sub-core grid configuration specifically for top-k operations
+        start_core: Starting core coordinate for sampling operations
+        num_gather_links: Number of links to use for all-gather operations (optional)
+        sampling_memory_config: Memory configuration for sampling tensors (optional)
+        k, p, temp: Initial sampling parameters (tensors of size max_batch_size)
+
+    Note:
+        Persistent buffers are used automatically when CCL supports line_all_gather (llama3_70b_galaxy),
+        and gracefully degrades to regular tt_all_gather for simpler CCL implementations (tt-transformers).
```
Comment (mtairum) [inline models/common/tt_sampling.py]: `gracefully degrades` 😅😅 🤔 🤔 ??

using persistent buffers vs. not using is no a degradation, we just let the CCL API deal with memory allocation.  Please update this comment.

File: models/common/tt_sampling.py [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+        else:
+            self.sampling_memory_config = ttnn.DRAM_MEMORY_CONFIG
+
+        # Set defaults for sampling parameters if not provided
+        if k is None:
+            k = torch.ones(self.max_batch_size)
+        if p is None:
+            p = torch.zeros(self.max_batch_size)
+        if temp is None:
+            temp = torch.ones(self.max_batch_size)
```
Comment (mtairum) [inline models/common/tt_sampling.py:93]: Are the defaults then:
K = 1
P = 0
Temp = 1

If P=0, we use argmax then?

File: models/common/tt_sampling.py [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+from models.common.lightweightmodule import LightweightModule
+
+
+class TTSampling(LightweightModule):
+    """
+    On-device sampling module supporting top-k, top-p, and temperature-based sampling.
+
+    This class implements high-performance on-device sampling that can work across different
+    model implementations by accepting configuration parameters rather than assuming specific
+    args structures.
```
Comment (mtairum) [inline models/common/tt_sampling.py:23]: Maybe add a line explaining how sampling is working on multi-device.
That we partition the vocab size across devices, etc.

File: models/common/tt_sampling.py [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+            return self.tt_ccl.line_all_gather(
+                tensor,
+                dim=dim,
+                cluster_axis=cluster_axis,
+                memory_config=memory_config,
+                num_links=num_links,
+                buffer_key=buffer_key,
+            )
+        else:
+            # Use tt_all_gather
```
Comment (mtairum) [inline models/common/tt_sampling.py:174]: We should move from this experimental op to the main API op.

For example All-gather should now be just `ttnn.all_gather` (https://github.com/tenstorrent/tt-metal/pull/30414)
Same for Reduce scatter - `ttnn.reduce_scatter` (https://github.com/tenstorrent/tt-metal/pull/29985)

Can you try the update APIs and report back?

This will save us coming back to this code and update in later on.

File: models/common/tt_sampling.py [hunk 11]
```diff
@@ -0,0 +1,380 @@
...
+            seed: Random seed for sampling
+            tt_out_tok: Optional output tensor to write results to
+
+        Returns:
+            Sampled token indices tensor
+        """
+        # Warmup run to ensure all operations are properly initialized
+        if self.warmup_done is False:
+            self.warmup_done = True
+            self.forward(x, seed=42, tt_out_tok=tt_out_tok)
```
Comment (mtairum) [inline models/common/tt_sampling.py:236]: Why is warmup needed?
I understand what's in the comment, but why should we need to warmup a sampling block? 
Maybe add more info to this comment.

File: models/common/tt_sampling.py [hunk 12]
```diff
@@ -0,0 +1,380 @@
...
+
+        # Convert to bfloat16 for top-k operations
+        x_bf16 = ttnn.typecast(x, dtype=ttnn.bfloat16, sub_core_grids=self.sub_core_grids)
```
Comment (mtairum) [inline models/common/tt_sampling.py:239]: Should we verify if x is bfp16 first? Or is this a no-op if converting bfp16 -> bfp16?

File: models/common/tt_sampling.py [hunk 13]
```diff
@@ -0,0 +1,380 @@
...
+                dtype=ttnn.bfloat16,
+            )
+            topk_values_gathered_bf16_interleaved = ttnn.to_memory_config(
+                topk_values_gathered_bf16, memory_config=ttnn.DRAM_MEMORY_CONFIG
+            )
+            ttnn.deallocate(topk_values_gathered_bf16)
+        else:
+            topk_values_gathered_bf16_interleaved = topk_values_gathered
+        # Gather top-k indices across all devices
+
```
Comment (mtairum) [inline models/common/tt_sampling.py]: This comments is decoupled from the function below. Aka, move it one line down.

File: models/common/tt_sampling.py [hunk 14]
```diff
@@ -0,0 +1,380 @@
...
+        )
+        ttnn.deallocate(topk_indices)
+
+        # Convert indices to appropriate data types
+        topk_indices_gathered_uint32 = ttnn.typecast(
+            topk_indices_gathered, dtype=ttnn.uint32, sub_core_grids=self.sub_core_grids
+        )
+        topk_indices_gathered_int32 = ttnn.typecast(
+            topk_indices_gathered_uint32, dtype=ttnn.int32, sub_core_grids=self.sub_core_grids
+        )
```
Comment (mtairum) [inline models/common/tt_sampling.py:319]: The all gather above cannot be in (u)int32, correct?
if so, leave a comment stating that, and that's the reason why we convert to (u)int32 after.

Comment (mtairum) [reply models/common/tt_sampling.py:319]: Also, we're typecasting uint16 -> uint32 -> int32.

I'm assuming skipping uint32 doesn't work? If so, should that be flagged as a bug?
Either way we need to add a comment here and open an issue and add the issue number as comment.

File: models/common/tt_sampling.py [hunk 15]
```diff
@@ -0,0 +1,380 @@
...
+            topk_values_gathered_bf16_interleaved,
+            topk_global_indices_interleaved_untilised,
+            k=self.k_tensor,
+            p=self.p_tensor,
+            temp=self.temp_tensor,
+            seed=seed,
+            sub_core_grids=ttnn.num_cores_to_corerangeset_in_subcoregrids(
+                self.start_core, self.max_batch_size, self.sub_core_grids, row_wise=True
+            )
+            if self.sub_core_grids != None
```
[the remark this answers is not in the document]
Comment (mtairum) [reply models/common/tt_sampling.py]: @sraizada-tt Apply this suggestion.

File: models/common/tt_sampling.py [hunk 16]
```diff
@@ -0,0 +1,380 @@
...
+    """
+    Format sampling parameters to a dictionary.
+    """
+    if not isinstance(sampling_params.temperature, List):
+        # convert all sampling_params to lists
+        update_dict = {field.name: [getattr(sampling_params, field.name)] for field in fields(sampling_params)}
+        sampling_params = replace(sampling_params, **update_dict)
+
+    # must pad sampling_params to max_batch_size
+    default_params = {"temp": 0.0, "p": 1.0, "k": 1.0}
```
[the remark this answers is not in the document]
Comment (mtairum) [reply models/common/tt_sampling.py]: @sraizada-tt apply this suggestion.

File: models/common/tt_sampling.py [hunk 17]
```diff
@@ -0,0 +1,380 @@
...
+        ("temp", "p", "k"), (sampling_params.temperature, sampling_params.top_p, sampling_params.top_k)
+    ):
+        current_len = len(tensor)
+        if current_len < target_len:
+            tensor.extend([default_params[name]] * (target_len - current_len))
+
+    # we must clamp top-p in range [0.0, 1.0)
+    # cannot rely on external SamplingParams to be clamped
+    TOP_P_MIN = 0.0
+    TOP_P_MAX = 0.99
```
Comment (mtairum) [inline models/common/tt_sampling.py]: Please add a comment referring why top-p-max is 0.99 (and link to issue if required).

**Review by mtairum — CHANGES_REQUESTED** [review]

File: models/common/tt_sampling.py (hunk 17 above)
Comment (bgoelTT) [reply models/common/tt_sampling.py]: @sraizada-tt I didn't need to clamp top-p in the Galaxy implementation, the negative tokens issue didn't surface in vLLM & after we did the seed warmup.

Did you reintroduce this because you ran into negative tokens again, even with the workaround? If we clamp then we should log this

File: models/tt_transformers/tt/model.py [hunk 18]
```diff
@@ -129,6 +132,19 @@ def __init__(
             max_columns_per_device=self.args.max_columns_per_device_lm_head,
         )
 
+        # Initialize on-device sampling if supported
+        # https://github.com/tenstorrent/tt-metal/issues/31134
+        # Blackhole devices do not support on-device sampling currently
+        # Sampling on device is supported only if each device has maximum logits size of 64*1024
+        sampling_splits = self.args.num_devices if list(self.mesh_device.shape) != [1, 1] else 2
+        self._supports_on_device_sampling = (
+            self.args.vocab_size // sampling_splits <= 64 * 1024 and "P" not in os.environ.get("MESH_DEVICE")
```
Comment (bgoelTT) [inline models/tt_transformers/tt/model.py]: Can we instead use a more programmatic API like `ttnn.device.is_blackhole()`? We should be moving away from the mesh device env var in my opinion.

This is also error prone

**Review by bgoelTT — CHANGES_REQUESTED** [review]

**Comment by mtairum** [comment]
TODO @sraizada-tt 
- Re-test sampling on BH CI
- Remove the fallback to argmax on BH
- Add an assert for sampling with non-linear temperatures (see https://github.com/tenstorrent/tt-metal/issues/25721), since we're still debugging that issue separately.

**Review by mtairum — APPROVED** [review]

**Review by bgoelTT — APPROVED** [review]

File: models/tt_transformers/tt/generator.py [hunk 19]
```diff
@@ -256,18 +257,27 @@ def decode_forward_text(
             sampling_params is None or sampling_params.temperature == 0
         ), "Currently only supporting greedy decoding (temperature=0) on device"
         argmax_on_device = sampling_params is not None and sampling_params.temperature == 0
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: Can the assert and argmax_on_device variable now be removed?

File: models/common/tt_sampling.py [hunk 20]
```diff
@@ -0,0 +1,412 @@
...
+    Format sampling parameters to a dictionary.
+    """
+    if not isinstance(sampling_params.temperature, List):
+        # convert all sampling_params to lists
+        update_dict = {field.name: [getattr(sampling_params, field.name)] for field in fields(sampling_params)}
+        sampling_params = replace(sampling_params, **update_dict)
+
+    # Must pad sampling_params to max_batch_size
+    default_params = {"temp": 0.0, "p": 1.0, "k": 1}
+    target_len = 32
```
Comment (skhorasganiTT) [inline models/common/tt_sampling.py]: Should this be referencing max batch size var instead?

File: models/tt_transformers/tt/generator.py [hunk 21]
```diff
@@ -256,18 +257,27 @@ def decode_forward_text(
...
         page_table = torch.chunk(page_table, self.data_parallel, 0) if page_table is not None else None
 
+        if sampling_on_device:
+            sampling_params = format_sampling_params(sampling_params)
+            for i in range(self.data_parallel):
+                self.model[i].tt_sampling.reset_params(
+                    k=sampling_params.top_k,
+                    p=sampling_params.top_p,
+                    temp=sampling_params.temperature,
+                )
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: This doesn't look like it's handling non-uniform sampling params across DPs. Shouldn't each DP model receive separate sampling params?

**Review by skhorasganiTT — CHANGES_REQUESTED** [review]

File: models/tt_transformers/tt/generator.py (hunk 21 above)
Comment (sraizada-tt) [reply models/tt_transformers/tt/generator.py]: I don't think vllm give us different sampling params per dp group that's why we have the sampling_params.temperature == 0 assert above instead of looping over sampling_params.[I]

File: models/tt_transformers/tt/generator.py (hunk 19 above)
Comment (sraizada-tt) [reply models/tt_transformers/tt/generator.py]: no, added comment

Comment (skhorasganiTT) [reply models/tt_transformers/tt/generator.py]: I saw you removed it?

File: models/tt_transformers/tt/generator.py (hunk 21 above)
Comment (skhorasganiTT) [reply models/tt_transformers/tt/generator.py]: vLLM gives sampling params for the global DP batch but here each DP model is using the same sampling params so I'm not sure how it works

File: models/tt_transformers/tt/generator.py (hunk 19 above)
Comment (sraizada-tt) [reply models/tt_transformers/tt/generator.py]: yeah, the issue that was blocking this got merged today

File: models/tt_transformers/tt/generator.py [hunk 22]
```diff
@@ -41,6 +42,21 @@ class SamplingParams:
...
+    chunk_size = len(lst) // n
+    remainder = len(lst) % n
+    chunks = []
+    start = 0
+    for i in range(n):
+        # Add 1 to chunk_size for the first 'remainder' chunks
+        end = start + chunk_size + (1 if i < remainder else 0)
+        chunks.append(list(lst[start:end]))  # Convert to list explicitly
+        start = end
+    return chunks
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: remainder stuff to be removed as remainder shouldn't be distributed between DPs and vLLM pads to max batch

**Review by skhorasganiTT — APPROVED** [review]

File: models/tt_transformers/tt/generator.py (hunk 21 above)
Comment (skhorasganiTT) [reply models/tt_transformers/tt/generator.py]: resolved by @sraizada-tt in newer commit
